homeassistant/components/madvr/remote.py

```diff
@@ -0,0 +1,176 @@
+        ]
+    )
+
+
+class MadvrRemote(CoordinatorEntity, RemoteEntity):
+    """Remote entity for the MadVR integration."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        coordinator: MadVRCoordinator,
+        madvr_client: Madvr,
+        entry_id: str,
+    ) -> None:
+        """Initialize the remote entity."""
+        super().__init__(coordinator)
+        self.madvr_client = madvr_client
+        self._attr_name = coordinator.name
+        self._attr_unique_id = f"{entry_id}_remote"
+        self.entry_id = hass.data[DOMAIN]["entry_id"]
+        self._attr_should_poll = False
+        self.tasks: list = []
+        self.connection_event = self.madvr_client.connection_event
+        self.command_queue: asyncio.Queue = asyncio.Queue()
+        self.stop_processing_commands = asyncio.Event()
+        self.madvr_client.set_update_callback(coordinator.handle_push_data)
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+        # open connection if not already connected
+        await super().async_added_to_hass()
+
+        # handle queue
+        task_queue = self.hass.loop.create_task(self.handle_queue())
+        self.tasks.append(task_queue)
+
+        task_notif = self.hass.loop.create_task(self.madvr_client.read_notifications())
+        self.tasks.append(task_notif)
+
+        task_hb = self.hass.loop.create_task(self.madvr_client.send_heartbeat())
+        self.tasks.append(task_hb)
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Run when removed."""
+        _LOGGER.debug("Removing from hass")
+        self.madvr_client.stop()
+        await self.madvr_client.close_connection()
+        await cancel_tasks(self.madvr_client)
+        for task in self.tasks:
+            if not task.done():
+                task.cancel()
+
+    @property
+    def is_on(self) -> bool:
+        """Return true if the device is on."""
+        return self.madvr_client.is_on
+
+    @property
+    def should_poll(self) -> bool:
+        """Poll."""
+        return False
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any]:
+        """Return extra state attributes."""
+        # Useful for making sensors
+        return self.madvr_client.msg_dict
+
+    async def handle_queue(self):
+        """Handle command queue."""
+        while True:
+            await self.connection_event.wait()
+            while (
+                not self.command_queue.empty()
+                and not self.stop_processing_commands.is_set()
+            ):
+                command = await self.command_queue.get()
+                _LOGGER.debug("sending queue command %s", command)
+                try:
+                    await self.madvr_client.send_command(command)
+                except ConnectionResetError:
+                    _LOGGER.warning("Envy was turned off manually")
+                    await self.async_turn_off()
```

Comment:
If the state is `OFF` please update the state and write the entities state. Further it seems not logical to turn off a device if it cannot be connected at this stage.

Reply:
updated this and will split out the functionality to disconnect and reset from "power off". The intention here is to mark the device as off